### PR TITLE
21987-Stdio-file-creation-should-be-able-to-be-deactivated

### PR DESCRIPTION
**Pull request by VincentBlondeau**
The way to deactivate it is to write the file in memory.
https://pharo.fogbugz.com/f/cases/21987/Stdio-file-creation-should-be-able-to-be-deactivated

### CODE REVIEW

**Comment by VincentBlondeau** [comment]
https://pharo.fogbugz.com/f/cases/21987/Stdio-file-creation-should-be-able-to-be-deactivated

**Comment by jecisc** [comment]
To keep a trace: I talked to Vincent and we think a better approach would be to use a NilStream instead of a memory stream. It would reduce the number of operations and ram consumption.

File: src/Files/Stdio.class.st [hunk 1]
```diff
@@ -29,6 +30,34 @@ Stdio class >> cleanStdioHandles [
...
+				do:
+					[ "Nothing, we do not have the rights to write and cannot suppose a dependency to the class MemoryFileSystemFile" ] ]
+]
+
+{ #category : #accessing }
+Stdio class >> createWriteStreamBlock [
+	"Can be changed to MemoryFileSystemFile in order to not write on the disk"
+
+	^ CreateWriteStreamBlock
+		ifNil: [ CreateWriteStreamBlock := [ :aName | (File named: aName) asString writeStream ] ]
```
Comment (akgrant43) [inline src/Files/Stdio.class.st]: > (File named: aName) asString

Doesn't this always return the string 'a File'?

**Comment by VincentBlondeau** [comment]
@akgrant It should be better! Thx :)

**Comment by akgrant43** [comment]
:-)

It looks better, but I haven't tested the patch.  I'm focused on trying to get FileAttributesPlugin changes integrated, and just found out that the Windows file times aren't working reliably. :-(

Cheers,
Alistair

**Comment by VincentBlondeau** [comment]
@akgrant43 if you need help to test the patch, I could help you

Cheers,
Vincent

**Comment by jecisc** [comment]
@VincentBlondeau 

IIUC this PR allow giving a block to get the streams on which write the stdout/in/err. Thus we now can use a block returning a NullStream instead of creating files? Am I correct?

In that case, I would like to have a method to directly disable the file creation.

For example

```Smalltalk
noLogInFiles
  self createWriteStreamBlock: [ NullStream new ]
```

Because not everyone knows that there exists a NullStream.

Also I would use #cull: instead of #value: because the user does not always need the parameter of the block.

**Comment by akgrant43** [comment]
Hi Vincent,

Sorry for the slow reply.  I've had very little time recently (unfortunately) and this dropped off my list.

Adding to what Cyril said:  

I'd also have a method to create the block for a memory stream (maybe #useNullStreams, #useMemoryStreams and #useDefaultStreams to put it back to the default).  One of the method comments suggests using MemoryFileSystemFile, but File can't simply be substituted with MemoryFileSystemFile since MemoryFileSystemFile doesn't understand #writeStream.

Also, the issue title suggests that it should always be possible to deactivate the file creation, and from Cyril's comments above, I suspect this is how he has interpreted it.  But the implementation only uses the CreateWriteStreamBlock if Pharo is not attached to a terminal.

This probably needs to be carefully commented to describe the intended behaviour under each of the possible scenarios, i.e. for each of stdin, stderr and stdout:

- the stream is attached to a terminal (default on posix platforms, PharoConsole.exe on Windows)
- the stream is redirected to a normal file, e.g. pharo pharo.image > output.txt
- the stream is redirected to a pipe, e.g. pharo pharo.image | tee output.txt

Cheers,
Alistair

**Comment by jecisc** [comment]
I like your methods names Alistair. Definitely better than the one I found.

**Comment by akgrant43** [comment]
> I like your methods names Alistair. Definitely better than the one I found.

Thanks, Cyril.  And if the <script> pragma is added we can easily configure the streams from the system browser. :-)

**Review by jecisc — APPROVED** [review]

**Comment by akgrant43** [comment]
Cyril has added his stamp of approval, so it's fine by me...

**Pull request merged**